boot/seal.go

```diff
@@ -93,8 +106,46 @@ func sealKeyToModeenv(key, saveKey secboot.EncryptionKey, model *asserts.Model,
 	return sealKeyToModeenvUsingSecboot(key, saveKey, model, modeenv)
 }
 
+func runKeySealRequests(key secboot.EncryptionKey) []secboot.SealKeyRequest {
+	return []secboot.SealKeyRequest{
+		{
+			Key:     key,
+			KeyFile: filepath.Join(InitramfsBootEncryptionKeyDir, "ubuntu-data.sealed-key"),
+		},
+	}
+}
+
+func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealKeyRequest {
+	return []secboot.SealKeyRequest{
+		{
+			Key:     key,
+			KeyFile: filepath.Join(InitramfsSeedEncryptionKeyDir, "ubuntu-data.recovery.sealed-key"),
+		},
+		{
+			Key:     saveKey,
+			KeyFile: filepath.Join(InitramfsSeedEncryptionKeyDir, "ubuntu-save.recovery.sealed-key"),
+		},
+	}
+}
+
 func sealKeyToModeenvUsingFdeSetupHook(key, saveKey secboot.EncryptionKey, model *asserts.Model, modeenv *Modeenv) error {
-	return fmt.Errorf("cannot use fde-setup hook yet")
+	// TODO: support full boot chains
+
+	for _, skr := range append(runKeySealRequests(key), fallbackKeySealRequests(key, saveKey)...) {
+		params := &FDESetupHookParams{
+			Key:     skr.Key,
+			KeyName: filepath.Base(skr.KeyFile),
+			Model:   model,
+		}
+		sealedKey, err := RunFDESetupHook("initial-setup", params)
+		if err != nil {
+			return fmt.Errorf("cannot run fde-setup hook for %s: %v", params.KeyName, err)
+		}
+		if err := osutil.AtomicWriteFile(filepath.Join(skr.KeyFile), sealedKey, 0600, 0); err != nil {
+			return fmt.Errorf("cannot store %s key: %v", params.KeyName, err)
+		}
+	}
```

Comment:
this should create the  sealed-keys stamp file I think, also maybe we should now write into it   something like "tpm" vs "fde-setup-hook" ?  they decision not to reseal should be taken based on calling features, not that file